internal/auth/authenticator.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2024 EngFlow Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"golang.org/x/oauth2"
+
+	"github.com/EngFlow/auth/internal/autherr"
+	"github.com/EngFlow/auth/internal/browser"
+)
+
+var errUnexpectedHTML = errors.New("request to JSON API returned HTML unexpectedly")
+
+type Backend interface {
+	Authenticate(ctx context.Context, host *url.URL) (*oauth2.Token, error)
+}
+
+type DeviceCode struct {
+	browserOpener browser.Opener
+	clientID      string
+	scopes        []string
+	httpTransport http.RoundTripper
```

Comment:
It doesn't seem like this can be initialized by the constructor. What's it for?

Reply:
Added a comment here, and explicitly set it to nil in the constructor so that it looks more intentional.